pkg/deploy/gateway.go

```diff
@@ -0,0 +1,446 @@
+	}
+
+	roleBinding := rbac.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GatewayServiceName,
+			Namespace: instance.Namespace,
+		},
+	}
+	if err := delete(clusterAPI, &roleBinding); err == nil {
+		return err
+	}
+
+	role := rbac.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GatewayServiceName,
+			Namespace: instance.Namespace,
+		},
+	}
+	if err := delete(clusterAPI, &role); err == nil {
+		return err
+	}
+
+	sa := corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GatewayServiceName,
+			Namespace: instance.Namespace,
+		},
+	}
+	if err := delete(clusterAPI, &sa); err == nil {
+		return err
+	}
+
+	return nil
+}
+
+// sync syncs the bluprint to the cluster in a generic (as much as Go allows) manner.
+func sync(instance *orgv1.CheCluster, clusterAPI ClusterAPI, blueprint metav1.Object, diffOpts cmp.Option) error {
+	blueprintObject, ok := blueprint.(runtime.Object)
+	if !ok {
+		return fmt.Errorf("object %T is not a runtime.Object. Cannot sync it", blueprint)
+	}
+
+	key := client.ObjectKey{Name: blueprint.GetName(), Namespace: blueprint.GetNamespace()}
+
+	actual := blueprintObject.DeepCopyObject()
+
+	if getErr := clusterAPI.Client.Get(context.TODO(), key, actual); getErr != nil {
+		if statusErr, ok := getErr.(*errors.StatusError); !ok || statusErr.Status().Reason != metav1.StatusReasonNotFound {
+			return getErr
+		}
+		actual = nil
+	}
+
+	kind := blueprintObject.GetObjectKind().GroupVersionKind().Kind
+
+	if actual == nil {
+		logrus.Infof("Creating a new object: %s, name %s", kind, blueprint.GetName())
+		err := clusterAPI.Client.Create(context.TODO(), blueprintObject)
+		if err != nil {
+			return err
+		}
+	} else {
+		actualMeta := actual.(metav1.Object)
+
+		diff := cmp.Diff(actual, blueprint, diffOpts)
+		if len(diff) > 0 {
+			logrus.Infof("Updating existing object: %s, name: %s", kind, actualMeta.GetName())
+			fmt.Printf("Difference:\n%s", diff)
+
+			err := clusterAPI.Client.Delete(context.TODO(), actual)
+			if err != nil {
+				return err
+			}
+
+			err = controllerutil.SetControllerReference(instance, blueprint, clusterAPI.Scheme)
+			if err != nil {
+				return err
+			}
+
+			err = clusterAPI.Client.Create(context.TODO(), blueprintObject)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func delete(clusterAPI ClusterAPI, obj metav1.Object) error {
+	key := client.ObjectKey{Name: obj.GetName(), Namespace: obj.GetNamespace()}
+	ro := obj.(runtime.Object)
+	if getErr := clusterAPI.Client.Get(context.TODO(), key, ro); getErr == nil {
+		if err := clusterAPI.Client.Delete(context.TODO(), ro); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// below functions declare the desired states of the various objects required for the gateway
+
+func getGatewayServerConfigSpec(instance *orgv1.CheCluster) corev1.ConfigMap {
+	data := `
+http:
+  routers:
+    che-server:
+      rule: "PathPrefix(` + "`/`" + `)"
+      service: che-server
+      priority: 100
```

Comment:
IMHO che-server should have the lowest priority. We're using `priority: 100` for workspaces so this would not work https://github.com/eclipse/che/blob/master/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/external/TraefikGatewayRouteConfigGenerator.java#L185